consul/query_endpoint.go

```diff
@@ -0,0 +1,454 @@
+			return err
+		}
+	}
+
+	// Try to locate the query.
+	state := q.srv.fsm.State()
+	_, query, err := state.QueryLookup(args.QueryIDOrName)
+	if err != nil {
+		return err
+	}
+	if query == nil {
+		return ErrQueryNotFound
+	}
+
+	// Execute the query for the local DC.
+	if err := q.execute(query, reply); err != nil {
+		return err
+	}
+
+	// Shuffle the results in case coordinates are not available if they
+	// requested an RTT sort.
+	reply.Nodes.Shuffle()
+	if query.Service.Sort == structs.QueryOrderSort {
+		if err := q.srv.sortNodesByDistanceFrom(args.Source, reply.Nodes); err != nil {
+			return err
+		}
+	}
+
+	// In the happy path where we found some healthy nodes we go with that
+	// and bail out. Otherwise, we fail over and try remote DCs, as allowed
+	// by the query setup.
+	if len(reply.Nodes) == 0 {
+		wrapper := &queryServerWrapper{q.srv}
+		if err := queryFailover(wrapper, query, args, reply); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ExecuteRemote is used when a local node doesn't have any instances of a
+// service available and needs to probe remote DCs. This sends the full query
+// over since the remote side won't have it in its state store, and this doesn't
+// do the failover logic since that's already being run on the originating DC.
+// We don't want things to fan out further than one level.
+func (q *Query) ExecuteRemote(args *structs.QueryExecuteRemoteRequest,
+	reply *structs.QueryExecuteResponse) error {
+	if done, err := q.srv.forward("Query.ExecuteRemote", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"consul", "query", "execute_remote"}, time.Now())
+
+	// We have to do this ourselves since we are not doing a blocking RPC.
+	if args.RequireConsistent {
+		if err := q.srv.consistentRead(); err != nil {
+			return err
+		}
+	}
+
+	// Run the query locally to see what we can find.
+	if err := q.execute(&args.Query, reply); err != nil {
+		return err
+	}
+
+	// We don't bother trying to do an RTT sort here since we are by
+	// definition in another DC. We just shuffle to make sure that we
+	// balance the load across the results.
+	reply.Nodes.Shuffle()
+
+	return nil
+}
+
+// execute runs a prepared query in the local DC without any failover. We don't
+// apply any sorting options at this level - it should be done up above.
+func (q *Query) execute(query *structs.PreparedQuery, reply *structs.QueryExecuteResponse) error {
+	state := q.srv.fsm.State()
+	_, nodes, err := state.CheckServiceNodes(query.Service.Service)
+	if err != nil {
+		return err
+	}
+
+	// This is kind of a paranoia ACL check, in case something changed with
+	// the token from the time the query was registered. Note that we use
+	// the token stored with the query, NOT the passed-in one, which is
+	// critical to how queries work (the query becomes a proxy for a lookup
+	// using the ACL it was created with).
+	if err := q.srv.filterACL(query.Token, nodes); err != nil {
```

Comment:
Need to do a better check here to see if the nodes got filtered due to ACLs. If so, there's no point in forwarding on to other DCs.